File: python-ref/gvibu_ref/commands/which_cmd.py
"""which: locate a command by searching PATH."""

import os
import sys


def run(args: list[str]) -> int:
    if not args:
        return 1

    path = os.environ.get("PATH", "")
    directories = [d for d in path.split(":") if d]
    found = False

    for name in args:
        for directory in directories:
            full_path = os.path.join(directory, name)
            if os.path.isfile(full_path) and os.access(full_path, os.X_OK):
                print(full_path)
                found = True
                break

    return 0 if found else 1

File: rust/src/commands/which_cmd.rs
/// which: locate a command by searching PATH.
use std::env;
use std::fs;
use std::os::unix::fs::PermissionsExt;

pub fn run(args: &[String]) -> i32 {
    if args.is_empty() {
        return 1;
    }

    let path = env::var("PATH").unwrap_or_default();
    let directories: Vec<&str> = path.split(':').filter(|d| !d.is_empty()).collect();

    let mut found = false;

    for name in args {
        for directory in &directories {
            let full_path = format!("{}/{}", directory, name);
            if let Ok(metadata) = fs::metadata(&full_path) {
                if metadata.is_file() {
                    let perms = metadata.permissions();
                    // Check if executable by owner
                    if perms.mode() & 0o111 != 0 {
                        println!("{}", full_path);
                        found = true;
                        break;
                    }
                }
            }
        }
    }

    if found { 0 } else { 1 }
}
